File: Recursion/Lecture 14/Fraz_recursion_14.java
import java.util.ArrayList;

class login {
    void help(int i, int k, int sumTillNow, int n, ArrayList<Integer> subSet, ArrayList<ArrayList<Integer>> ans) {

        if (sumTillNow > n)
            return;
        if (k == 0) {
            if (sumTillNow == n) {
                ans.add(subSet);
            }
            return;
        }
        // if( k < 0) return;
        // if(sumTillNow == n) {
        // if( k == 0) {
        // ans.push_back(subSet);
        // }
        // return;
        // }

        if (i == 10)
            return;
        // pick
        sumTillNow += i;
        subSet.add(i);
        help(i + 1, k - 1, sumTillNow, n, subSet, ans);
        sumTillNow -= i;
        subSet.remove(subSet.size() - 1);

        // ignore
        help(i + 1, k, sumTillNow, n, subSet, ans);

    }

    ArrayList<ArrayList<Integer>> combinationSum3(int k, int n) {
        ArrayList<Integer> subSet = new ArrayList<>();
        ArrayList<ArrayList<Integer>> ans = new ArrayList<>();
        help(1, k, 0, n, subSet, ans);
        return ans;
    }
}

File: Recursion/Lecture 14/Fraz_recursion_14.cpp
#include <bits/stdc++.h>
using namespace std;

void help(int i, int k, int sumTillNow, int n, vector<int> &subSet, vector<vector<int>> &ans) {
    
    if(k == 0) {
        if(sumTillNow == n) {
            ans.push_back(subSet);
        }
        return;
    }
    if(i == 10) return;
    sumTillNow += i;
    subSet.push_back(i);
    help(i + 1, k - 1, sumTillNow, n, subSet, ans);
    sumTillNow -= i;
    subSet.pop_back();
    help(i + 1, k, sumTillNow, n, subSet, ans); 

}

vector<vector<int>> combinationSum3(int k, int n) {
    vector<int> subSet;
    vector<vector<int>> ans;
    help(1, k, 0, n, subSet, ans);
    return ans;
}
